File: src/phase-loop.ts
import fs from 'node:fs'
import path from 'node:path'
import readline from 'node:readline'
import {
  type Agent,
  type PlanContext,
  type SharedConfig,
  type Phase,
  LLMPhase,
  ScriptPhase,
  isScriptPhase,
} from './types.js'
import { composePrompt } from './prompt-composer.js'
import { shouldDream, updateDreamBaseline } from './dream.js'
import { gitCommitPlan, gitSaveWorkBaseline } from './git.js'
import { dispatchSubagents } from './subagent-dispatch.js'

function readPhase(planDir: string): Phase {
  const phasePath = path.join(planDir, 'phase.md')
  return fs.readFileSync(phasePath, 'utf-8').trim() as Phase
}

function writePhase(planDir: string, phase: Phase): void {
  fs.writeFileSync(path.join(planDir, 'phase.md'), phase)
}

function planName(planDir: string): string {
  return path.basename(planDir)
}

async function askContinue(): Promise<boolean> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })
  return new Promise(resolve => {
    rl.question('\nContinue to next phase? [Y/n] ', answer => {
      rl.close()
      resolve(answer.trim().toLowerCase() !== 'n')
    })
  })
}

async function handleScriptPhase(
  phase: ScriptPhase,
  planDir: string
): Promise<boolean> {
  const name = planName(planDir)

  switch (phase) {
    case ScriptPhase.GitCommitWork:
      gitCommitPlan(planDir, name, 'work')
      writePhase(planDir, LLMPhase.Reflect)
      return askContinue()

    case ScriptPhase.GitCommitReflect:
      gitCommitPlan(planDir, name, 'reflect')
      writePhase(planDir, LLMPhase.Dream)
      return true

    case ScriptPhase.GitCommitDream:
      gitCommitPlan(planDir, name, 'dream')
      writePhase(planDir, LLMPhase.Triage)
      return true

    case ScriptPhase.GitCommitTriage:
      gitCommitPlan(planDir, name, 'triage')
      writePhase(planDir, LLMPhase.Work)
      return askContinue()
  }
}

export async function phaseLoop(
  agent: Agent,
  ctx: PlanContext,
  config: SharedConfig,
  projectRoot: string
): Promise<void> {
  const tokens = agent.tokens()

  if (agent.setup) {
    await agent.setup(ctx)
  }

  while (true) {
    const phase = readPhase(ctx.planDir)
    console.log(`\n▶ Phase: ${phase}`)

    if (isScriptPhase(phase)) {
      const shouldContinue = await handleScriptPhase(phase, ctx.planDir)
      if (!shouldContinue) {
        console.log('Exiting.')
        return
      }
      continue
    }

    // Pre-work: save baseline for analyse-work diff
    if (phase === LLMPhase.Work) {
      gitSaveWorkBaseline(ctx.planDir)
      fs.rmSync(path.join(ctx.planDir, 'latest-session.md'), { force: true })
    }

    const prompt = composePrompt(projectRoot, phase, ctx, tokens)

    if (phase === LLMPhase.Work) {
      await agent.invokeInteractive(prompt, ctx)
    } else {
      await agent.invokeHeadless(prompt, ctx, phase)
    }

    // Post-phase: check phase advanced
    const newPhase = readPhase(ctx.planDir)
    if (newPhase === phase) {
      console.error(`⚠ Phase did not advance from ${phase}. Stopping.`)
      return
    }

    // Dream trigger: after reflect commits, check if dream is needed
    if (newPhase === LLMPhase.Dream || readPhase(ctx.planDir) === LLMPhase.Dream) {
      if (!shouldDream(ctx.planDir, config.headroom)) {
        console.log('  ⏭ Dream skipped (memory within headroom)')
        writePhase(ctx.planDir, ScriptPhase.GitCommitDream)
      }
    }

    // After dream phase completes, update baseline
    if (phase === LLMPhase.Dream) {
      updateDreamBaseline(ctx.planDir)
    }

    // After triage, dispatch subagents
    if (phase === LLMPhase.Triage) {
      await dispatchSubagents(agent, ctx.planDir)
    }
  }
}

File: raveloop-cli/src/phase_loop.rs
use std::fs;
use std::path::Path;
use std::sync::Arc;

use anyhow::{Context, Result};

use crate::agent::Agent;
use crate::dream::{should_dream, update_dream_baseline};
use crate::format::phase_info;
use crate::git::{git_commit_plan, git_save_work_baseline};
use crate::prompt::compose_prompt;
use crate::subagent::dispatch_subagents;
use crate::types::*;
use crate::ui::UI;

const HR: &str = "────────────────────────────────────────────────────";

fn read_phase(plan_dir: &Path) -> Result<Phase> {
    let content = fs::read_to_string(plan_dir.join("phase.md"))
        .context("Failed to read phase.md")?;
    Phase::parse(content.trim())
        .with_context(|| format!("Unknown phase: {}", content.trim()))
}

fn write_phase(plan_dir: &Path, phase: Phase) {
    let _ = fs::write(plan_dir.join("phase.md"), phase.to_string());
}

fn plan_name(plan_dir: &Path) -> String {
    plan_dir.file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_default()
}

fn log_phase_header(ui: &UI, phase: LlmPhase, plan: &str) {
    let info = phase_info(phase);
    ui.log(&format!("\n{HR}"));
    ui.log(&format!("  ◆  {}  ·  {plan}", info.label));
    ui.log(&format!("  {}", info.description));
    ui.log(HR);
}

fn log_commit(ui: &UI, phase_name: &str, plan: &str, result: &crate::git::CommitResult) {
    if result.committed {
        let first_line = result.message.lines().next().unwrap_or("");
        ui.log(&format!("\n  ⚙  COMMIT · {phase_name}  ·  {plan}  ·  {first_line}"));
    } else {
        ui.log(&format!("\n  ⚙  COMMIT · {phase_name}  ·  {plan}  ·  nothing to commit"));
    }
}

async fn handle_script_phase(
    phase: ScriptPhase,
    plan_dir: &Path,
    headroom: usize,
    ui: &UI,
) -> Result<bool> {
    let name = plan_name(plan_dir);

    match phase {
        ScriptPhase::GitCommitWork => {
            let result = git_commit_plan(plan_dir, &name, "work")?;
            log_commit(ui, "work", &name, &result);
            write_phase(plan_dir, Phase::Llm(LlmPhase::Reflect));
            Ok(ui.confirm("Proceed to reflect phase?").await)
        }
        ScriptPhase::GitCommitReflect => {
            let result = git_commit_plan(plan_dir, &name, "reflect")?;
            log_commit(ui, "reflect", &name, &result);
            if should_dream(plan_dir, headroom) {
                write_phase(plan_dir, Phase::Llm(LlmPhase::Dream));
            } else {
                ui.log("  ⏭  Dream skipped (memory within headroom)");
                write_phase(plan_dir, Phase::Script(ScriptPhase::GitCommitDream));
            }
            Ok(true)
        }
        ScriptPhase::GitCommitDream => {
            let result = git_commit_plan(plan_dir, &name, "dream")?;
            log_commit(ui, "dream", &name, &result);
            write_phase(plan_dir, Phase::Llm(LlmPhase::Triage));
            Ok(true)
        }
        ScriptPhase::GitCommitTriage => {
            let result = git_commit_plan(plan_dir, &name, "triage")?;
            log_commit(ui, "triage", &name, &result);
            write_phase(plan_dir, Phase::Llm(LlmPhase::Work));
            Ok(ui.confirm("Proceed to next work phase?").await)
        }
    }
}

pub async fn phase_loop(
    agent: Arc<dyn Agent>,
    ctx: &PlanContext,
    config: &SharedConfig,
    ui: &UI,
) -> Result<()> {
    let tokens = agent.tokens();
    let plan_dir = Path::new(&ctx.plan_dir);
    let config_root = Path::new(&ctx.config_root);
    let name = plan_name(plan_dir);

    if let Err(e) = agent.setup(ctx).await {
        ui.log(&format!("  ✗  Setup failed: {e}"));
    }

    loop {
        let phase = read_phase(plan_dir)?;

        match phase {
            Phase::Script(sp) => {
                if !handle_script_phase(sp, plan_dir, config.headroom, ui).await? {
                    ui.log("\nExiting.");
                    return Ok(());
                }
                continue;
            }
            Phase::Llm(lp) => {
                let agent_id = "main";
                log_phase_header(ui, lp, &name);

                ui.set_status(StatusInfo {
                    project: Path::new(&ctx.project_dir)
                        .file_name()
                        .map(|n| n.to_string_lossy().to_string())
                        .unwrap_or_default(),
                    plan: name.clone(),
                    phase: lp.as_str().to_string(),
                    agent: config.agent.clone(),
                    cycle: None,
                });

                if lp == LlmPhase::Work {
                    git_save_work_baseline(plan_dir);
                    let _ = fs::remove_file(plan_dir.join("latest-session.md"));
                }

                let prompt = compose_prompt(config_root, lp, ctx, &tokens)?;
                let tx = ui.sender();

                ui.register_agent(agent_id, &format!("  ◆  {}  ·  {name}", phase_info(lp).label));

                if lp == LlmPhase::Work {
                    ui.suspend();
                    agent.invoke_interactive(&prompt, ctx).await?;
                    ui.resume();
                } else {
                    agent.invoke_headless(&prompt, ctx, lp, agent_id, tx).await?;
                }

                let new_phase = read_phase(plan_dir)?;
                if new_phase == phase {
                    ui.log(&format!("\n  ✗  Phase did not advance from {phase}. Stopping."));
                    return Ok(());
                }

                if lp == LlmPhase::Dream {
                    update_dream_baseline(plan_dir);
                }

                if lp == LlmPhase::Triage {
                    dispatch_subagents(agent.clone(), plan_dir, ui).await?;
                }
            }
        }
    }
}
